precompiles/act/act.go

```diff
@@ -0,0 +1,162 @@
+// for the x/act precompile.
+func LoadABI() (abi.ABI, error) {
+	return evmoscmn.LoadABI(f, "abi.json")
+}
+
+func NewPrecompile(actkeeper actmodulekeeper.Keeper, er *common.EthEventsRegistry) (*Precompile, error) {
+	abi, err := LoadABI()
+	if err != nil {
+		return nil, err
+	}
+
+	p := Precompile{
+		Precompile: evmoscmn.Precompile{
+			ABI:                  abi,
+			KvGasConfig:          storetypes.KVGasConfig(),
+			TransientKVGasConfig: storetypes.TransientGasConfig(),
+		},
+		actmodulekeeper: actkeeper,
+		eventsRegistry:  er,
+	}
+
+	return &p, nil
+}
+
+// Address implements vm.PrecompiledContract.
+func (p *Precompile) Address() ethcmn.Address {
+	return ethcmn.HexToAddress(PrecompileAddress)
+}
+
+// RequiredGas implements vm.PrecompiledContract.
+// Subtle: this method shadows the method (Precompile).RequiredGas of Precompile.Precompile.
+func (p *Precompile) RequiredGas(input []byte) uint64 {
+	methodID := input[:4]
+
+	method, err := p.MethodById(methodID)
+	if err != nil {
+		// This should never happen since this method is going to fail during Run
+		return 0
+	}
+
+	return p.Precompile.RequiredGas(input, p.IsTransaction(method.Name))
+}
+
+// Run implements vm.PrecompiledContract.
+func (p *Precompile) Run(evm *vm.EVM, contract *vm.Contract, readonly bool) (bz []byte, err error) {
+	ctx, stateDB, _, method, initialGas, args, err := p.RunSetup(evm, contract, readonly, p.IsTransaction)
+	if err != nil {
+		return nil, err
+	}
+
+	// This handles any out of gas errors that may occur during the execution of a precompile tx or query.
+	// It avoids panics and returns the out of gas error so the EVM can continue gracefully.
+	defer evmoscmn.HandleGasError(ctx, contract, initialGas, &err)()
+
+	if err := stateDB.Commit(); err != nil {
+		return nil, err
+	}
+
+	switch method.Name {
+	// transactions
+	case CheckActionMethod:
+		bz, err = p.CheckActionMethod(ctx, evm.Origin, stateDB, method, args)
+	case NewTemplateMethod:
+		bz, err = p.NewTemplateMethod(ctx, evm.Origin, stateDB, method, args)
+	case RevokeActionMethod:
+		bz, err = p.RevokeActionMethod(ctx, evm.Origin, stateDB, method, args)
+	case UpdateTemplateMethod:
+		bz, err = p.UpdateTemplateMethod(ctx, evm.Origin, stateDB, method, args)
+	case VoteForActionMethod:
+		bz, err = p.VoteForActionMethod(ctx, evm.Origin, stateDB, method, args)
+	// queries
+	case ActionsQuery:
+		bz, err = p.ActionsQuery(ctx, contract, method, args)
+	case ActionByIdQuery:
+		bz, err = p.ActionByIdQuery(ctx, contract, method, args)
+	case ActionsByAddressQuery:
+		bz, err = p.ActionsByAddressQuery(ctx, contract, method, args)
+	case TemplatesQuery:
+		bz, err = p.TemplatesQuery(ctx, contract, method, args)
+	case TemplateByIdQuery:
+		bz, err = p.TemplateByIdQuery(ctx, contract, method, args)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	cost := ctx.GasMeter().GasConsumed() - initialGas
+
+	if !contract.UseGas(cost) {
+		return nil, vm.ErrOutOfGas
+	}
+
```

Comment:
```suggestion

	if err := p.AddJournalEntries(stateDB, snapshot); err != nil {
		return nil, err
	}
```

i see this in the `staking` precompile from evmos, not sure what it does. Do we need it?

Reply:
https://github.com/evmos/evmos/blob/main/precompiles/staking/staking.go#L149-L151

Reply:
> This allows to revert the call changes within an evm tx

Looks like we need it